src/Data/GADT/Coerce.hs

```diff
@@ -0,0 +1,107 @@
+import Data.IORef           (IORef)
+import Data.STRef           (STRef)
+import Data.Type.Equality   ((:~:) (..))
+import Data.Type.Coercion
+
+import Unsafe.Coerce        (unsafeCoerce)
+
+import Data.GADT.Compare    (GEq, geq)
+
+#if MIN_VERSION_base(4,10,0)
+import           Data.Type.Equality ((:~~:) (..))
+import qualified Type.Reflection    as TR
+#endif
+
+#if __GLASGOW_HASKELL__ >= 810
+import Data.Kind (Type, Constraint)
+#endif
+
+-- |A class for type-contexts which contain enough information
+-- to (at least in some cases) decide the coercibility of types
+-- occurring within them.
+#if __GLASGOW_HASKELL__ >= 810
+type GCoercible :: (k -> Type) -> Constraint
+#endif
+class GCoercible f where
+    gcoercible :: f a -> f b -> Maybe (Coercion a b)
+
+-- |If 'f' has a 'GEq' instance, this function makes a suitable default
+-- implementation of 'gcoercible'.
+defaultGcoercible :: GEq f => f a -> f b -> Maybe (Coercion a b)
+defaultGcoercible x y = fmap repr $ geq x y
+
+instance GCoercible ((:~:) a) where
+    gcoercible = defaultGcoercible
+
+#if MIN_VERSION_base(4,10,0)
+instance GCoercible ((:~~:) a) where
+    gcoercible = defaultGcoercible
+
+instance GCoercible TR.TypeRep where
+    gcoercible = defaultGcoercible
+#endif
+
+instance (GCoercible a, GCoercible b) => GCoercible (Sum a b) where
+    gcoercible (InL x) (InL y) = gcoercible x y
+    gcoercible (InR x) (InR y) = gcoercible x y
+    gcoercible _ _ = Nothing
+
+instance (GCoercible a, GCoercible b) => GCoercible (Product a b) where
+    gcoercible (Pair x y) (Pair x' y') = do
+        Coercion <- gcoercible x x'
+        Coercion <- gcoercible y y'
+        return Coercion
+
+instance GCoercible IORef where
+    gcoercible x y =
+        if x == unsafeCoerce y
+        then Just $ unsafeCoerce $ repr Refl
+        else Nothing
+
+instance GCoercible (STRef s) where
+    gcoercible x y =
+        if x == unsafeCoerce y
+        then Just $ unsafeCoerce $ repr Refl
+        else Nothing
+
+-- This instance seems nice, but it's simply not right:
+--
+-- > instance GCoercible StableName where
+-- >     gcoercible sn1 sn2
+-- >         | sn1 == unsafeCoerce sn2
+-- >             = Just (unsafeCoerce Refl)
+-- >         | otherwise     = Nothing
+--
+-- Proof:
+--
+-- > x <- makeStableName id :: IO (StableName (Int -> Int))
+-- > y <- makeStableName id :: IO (StableName ((Int -> Int) -> Int -> Int))
+-- >
+-- > let Just boom = gcoercible x y
+-- >
+-- > Data.Type.Coercion.coerceWith boom (const 0) id 0
+-- > let "Illegal Instruction" = "QED."
+--
+-- The core of the problem is that 'makeStableName' only knows the closure it is
+-- passed to, not any type information. Together with the fact that the same
+-- closure has the same 'StableName' each time 'makeStableName' is called on it,
+-- there is potential for abuse when a closure can be given many incompatible
+-- types.
+--
+-- 'GCoericble' gets us closer than GEq, but the problem is Coercions state that
```

Comment:
There probably should be a class for heterogenous equality

```haskell
class (forall a. Eq (f a)) => HEq f where
  heq :: f a -> f b -> Bool
```

with examples like

```haskell
instance HEq Proxy where
  heq _ _ = True
```

and

```haskell
instance HEq StableName where
  heq = eqStableName
```

(and maybe for ordering as well), which would be a natural superclass for `GCoercible` (and `GEq`).

But that should wait for when I'm ready for QuantifiedConstraints as well

Reply:
Mmm that class feels insufficient canonical to me. It is possible to define a valid equivalence class that is completely unrelated to the `GEq` or `GCoercible` instances of a type.

That said,
```haskell
newtype HEq f = MkHeq { unHeq :: forall a b. f a -> f b -> Bool }
```
might be useful to have. Not sure whether it goes in this library but certainly it should be possible to get values of `HEq f` if one has `GEq` f` or `GCoercible f `.

Reply:
Compare

```haskell
class (forall a. Eq (f a)) => HEq f where
  heq :: f a -> f b -> Bool
```

and

```haskell
class (forall a. Eq a => Eq (f a)) => Eq1 f where
  liftEq :: (a -> b -> Bool) -> f a -> f b -> Bool
```

There's nothing non-canonical in `HEq`. It extends `Eq (f a)` into heterogenous setting when possible. `StableName`, `Proxy` etc.